packages/yew-router/src/service.rs

```diff
@@ -1,90 +1,162 @@
+                Some(move || {
+                    for subscriber in subscribers {
+                        subscriber.emit(route.clone());
+                    }
+                })
             } else {
-                path
+                None
             }
+        })
+    }
+
+    fn handle_error(error: T) {
+        // Whenever we fail to recognize a route, we redirect to the default one
+        history::replace(error.to_route());
+    }
+
+    fn update(route: Rc<Route>) {
+        match Self::with(|entry| Self::update_inner(entry, route)) {
+            Ok(None) => {}
+            Ok(Some(f)) => f(),
+            Err(error) => Self::handle_error(error),
         }
-        None => url,
-    };
-
-    history
-        .push_state_with_url(&JsValue::NULL, "", Some(&path))
-        .expect("push history");
-    let event = Event::new("popstate").unwrap();
-    yew::utils::window()
-        .dispatch_event(&event)
-        .expect("dispatch");
-}
+    }
 
-pub fn parse_query<T>() -> Result<T, serde_urlencoded::de::Error>
-where
-    T: for<'de> Deserialize<'de>,
-{
-    let query = yew::utils::document().location().unwrap().search().unwrap();
-    serde_urlencoded::from_str(query.strip_prefix("?").unwrap_or(""))
-}
+    fn current_route(entry: Entry<Self>) -> Result<Rc<T>, T> {
+        Ok(Self::try_insert(entry)?.last_route.clone())
+    }
+
+    fn register(entry: Entry<Self>, callback: Callback<Rc<T>>) -> Result<RouteListener<T>, T> {
+        Self::try_insert(entry)?.subscribers.push(callback.clone());
+        Ok(RouteListener(callback))
+    }
 
-pub fn current_route<R: Routable>() -> Option<R> {
-    R::current_route()
+    fn unregister(entry: Entry<Self>, callback: &Callback<Rc<T>>) {
+        if let Entry::Occupied(occ) = entry {
+            let state = occ.into_mut();
+            if let Some(index) = state
+                .subscribers
+                .iter()
+                .position(|subscriber| subscriber == callback)
+            {
+                state.subscribers.remove(index);
+            }
+        }
+    }
 }
+pub struct RouteListener<T: Routable>(Callback<Rc<T>>);
 
-/// Handle for the router's path event listener
-pub struct RouteListener {
-    // this exists so listener is dropped when handle is dropped
-    #[allow(dead_code)]
-    listener: EventListener,
+impl<T: Routable> Drop for RouteListener<T> {
+    fn drop(&mut self) {
+        RouterState::with(|entry| RouterState::unregister(entry, &self.0));
+    }
 }
 
-/// Adds a listener which is called when the current route is changed.
-///
-/// The callback receives `Option<R>` so it can handle the error case itself.
-pub fn attach_route_listener<R>(callback: Callback<Option<R>>) -> RouteListener
-where
-    R: Routable + 'static,
-{
-    let listener = EventListener::new(&yew::utils::window(), "popstate", move |_| {
-        callback.emit(current_route())
-    });
-
-    RouteListener { listener }
+pub fn dispatch<T: Routable>(action: RouterAction<T>) {
```

Comment:
Since there are functions that call this function, I think this should be `pub(crate)`.

Reply:
It is more useful for the `Mount` functionality, where you might need to adapt a `RouteAction<T>` to a `RouteAction<U>` and pass it on, I'll make it private in this PR.

Reply:
Yeah, it's better to keep it private in this PR.